### PR TITLE
Document graceful degradation rules and add tests

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- Documents graceful degradation rules for Python exception logging and traceback capture.
- Adds a comprehensive suite of tests validating graceful degradation across both exceptions and frames, including optional fields, missing attributes, type mismatches, and edge cases in locals extraction.

### Documentation
- Update docs/adr-001-python-exception-logging.md to include a dedicated section on Graceful degradation rules.
- Describe required vs optional extraction, the get_optional_attr helper, partial extraction of collections, and the design rationale and consequences.

### Tests
- rust_extension/src/traceback_capture_graceful_degradation_tests.rs: new tests covering graceful degradation for exception capture (non-iterable __notes__, non-string note elements, and other attribute degradation paths).
- rust_extension/src/traceback_frames_graceful_degradation_tests.rs: new tests covering frame extraction resilience, including:
  - frame_with_wrong_type_optional_field_degrades_to_none
  - frame_with_explicit_none_optional_fields_degrades_to_none
  - extract_locals_with_unrepr_able_value_skips_entry
  - extract_locals_with_non_mapping_locals_degrades_gracefully
  - extract_locals_with_non_string_keys_degrades_gracefully
- rust_extension/src/traceback_frames_tests.rs
  - frame_with_all_optional_fields_present
  - frame_with_missing_optional_fields
  (existing tests retained and continue to validate standard extraction paths.)
- rust_extension/src/test_utils/traceback_test_helpers.rs: updated to support new graceful degradation tests with additional assertion helpers.

### Implementation notes
- The codebase now includes graceful degradation test modules to validate resilience against malformed or partial traceback data, ensuring the logging system remains operable.

## Why
- Provides a robust, version-agnostic approach to extracting diagnostic data from Python exceptions.
- Ensures the logging system remains operational even with malformed or unusual traceback shapes.
- Improves resilience and stability of error reporting by gracefully degrading optional fields without failing captures.

## Testing plan
- [x] Run cargo test for rust_extension tests locally.
- [x] Validate optional field degradation paths for frames and exceptions.
- [x] Verify that missing attributes and non-repr-able values do not crash extraction logic.
- [x] Exercise additional graceful degradation paths via new tests for non-iterable/non-string notes, non-mapping locals, and non-string keys.

## Notes
- No public API changes; tests verify behavior aligns with the documented Graceful degradation rules.

🌿 Generated by [Terry](https://www.terragonlabs.com)

---
ℹ️ Tag @terragon-labs to ask questions and address PR feedback

📎 **Task**: https://www.terragonlabs.com/task/033d9608-3f82-490f-907b-d47f8e4972ca

## Summary by Sourcery

Document graceful degradation behavior for Python exception logging and add regression tests to validate frame and exception capture under malformed or partial traceback conditions.

Documentation:
- Describe required vs optional traceback/exception fields and the graceful degradation rules applied during extraction, including collection handling and rationale.

Tests:
- Add tests ensuring optional stack frame fields with wrong types or explicit None degrade to None and that locals entries with unrepr-able values are skipped while valid entries are preserved.
- Add tests verifying exception capture gracefully handles missing notes, empty args, and chained causes where notes and args cannot be recovered.

### CODE REVIEW

File: rust_extension/src/traceback_capture_graceful_degradation_tests.rs [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+        assert!(
+            cause.notes.is_empty(),
+            "chained exception notes should be empty (no instance access)"
+        );
+        assert!(
+            cause.args_repr.is_empty(),
+            "chained exception args_repr should be empty (no instance access)"
+        );
+    });
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply rust_extension/src/traceback_capture_graceful_degradation_tests.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: rust_extension/src/test_utils/traceback_test_helpers.rs [hunk 2]
```diff
@@ -275,3 +277,60 @@ pub fn populate_locals_dict_from_entries(locals_dict: &Bound<'_, PyDict>, entrie
...
+pub fn assert_frame_required_fields(
+    frame: &StackFrame,
+    filename: &str,
+    lineno: u32,
+    function: &str,
+) {
+    assert_eq!(frame.filename, filename, "filename should match");
+    assert_eq!(frame.lineno, lineno, "lineno should match");
+    assert_eq!(frame.function, function, "function should match");
+}
```
[the remark this answers is not in the document]
Comment (leynos) [reply rust_extension/src/test_utils/traceback_test_helpers.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: rust_extension/src/traceback_capture.rs [hunk 3]
```diff
@@ -11,26 +11,13 @@
 //! and [`capture_stack`] for `stack_info=True` support.
 
 use pyo3::prelude::*;
-use pyo3::types::{PyBool, PyDict, PyList, PyTuple};
+use pyo3::types::{PyBool, PyDict, PyList, PyString, PyTuple};
```
[the remark this answers is not in the document]
Comment (leynos) [reply rust_extension/src/traceback_capture.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: rust_extension/src/traceback_capture_graceful_degradation_tests.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (leynos) [reply rust_extension/src/traceback_capture_graceful_degradation_tests.rs]: @coderabbitai The test passes. The current code is correct:

1.  **`err.value(py)`** - This IS the correct modern PyO3 0.25 API. The method `value_bound()` does **not exist** in PyO3 0.25.1. The `value()` method already returns `&Bound<'py, PyBaseException>`, which is the "bound" form.
    
2.  **`capture_exception(py, &exc_value)`** - Already passing by reference as required by the function signature.
    
The reviewer's suggested change to use `value_bound(py)` would cause a compilation error because that method doesn't exist. The code is already correct and all tests pass.

**Comment by leynos** [comment]
@coderabbitai review

File: rust_extension/src/traceback_capture_graceful_degradation_tests.rs [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+        let exc = create_value_error(py, "test error");
+        let args_tuple = pyo3::types::PyTuple::new(
+            py,
+            [
+                "valid arg".into_pyobject(py).unwrap().into_any(),
+                bad_repr_instance,
+                42i32.into_pyobject(py).unwrap().into_any(),
+            ],
+        )
+        .expect("tuple creation should succeed");
```
[the remark this answers is not in the document]
Comment (leynos) [reply rust_extension/src/traceback_capture_graceful_degradation_tests.rs]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.
